### PR TITLE
[dv/tools] Updated Coverage flow for xcelium

### PR DESCRIPTION
**Pull request by rasmus-madsen**
Hi @sriyerg
so I added the ccf and the vRefine files
and I also put in a cmd so that each module can add its own refine file.

I didn't actually exclude anything yet. so the file is just a placeholder.

the ccf file should match what you have done for vcs more or less.
I didn't find the command for static signals. and I figured we cross that bridge once we run into a problem.

take a look and tell me what you think

### CODE REVIEW

File: hw/dv/tools/xcelium/xcelium.ccf [hunk 1]
```diff
@@ -0,0 +1,8 @@
+//only collect toggle coverage on dut
```
Comment (sriyerg) [inline hw/dv/tools/xcelium/xcelium.ccf:1]: Please add lowrisc license banner.

File: hw/dv/tools/xcelium/common_cov.vRefine [hunk 2]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="ISO-8859-1" standalone="no"?>
+<refinement-file-root>
+  <information comment-version="1" creation-time="Wed 10 Feb 2021 04:11:52 PST" creator="rasmus.madsen" csCheck="false" modified-by="rasmus.madsen" modify-time="Wed 10 Feb 2021 05:18:29 PST" save-ref-method="seq" tool-version="Cadence vManager20.03" rules-signature-c="1ead9ec7f5923387d479a9c435d026b0">
+    <ucm-files>
+      <ucm-file domain="icc" modelCheckSum="1468209578" path="/wdc/proj/cto/top/work/rasmus.madsen/opentitan/golden/hw/ip/aes/dv/scratch/tools_cov/aes-sim-xcelium/cov_merge/merged/icc_578319aa_69124526.ucm"></ucm-file>
```
Comment (weicaiyang) [inline hw/dv/tools/xcelium/common_cov.vRefine:5]: this is meant to be a common cov cfg file, but it contains your local path, not sure if we can remove it
does this work for other IPs?

File: hw/dv/tools/xcelium/common_cov.vRefine [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    </ucm-files>
+    <ccf-files>
+      <ccf-file content="//only collect toggle coverage on dut;deselect_coverage -toggle -instance tb.dut.*;// only collect toggle coverage on ports;set_toggle_portsonly;// enable toggle scoring of structs and multidim arrays + MDA structs;set_toggle_scoring -sv_mda -sv_mda_of_struct;// filter glitches;set_toggle_strobe 0ps;" path="/wdc/proj/cto/top/work/rasmus.madsen/opentitan/golden/hw/dv/tools/xcelium/xcelium.ccf"></ccf-file>
```
Comment (weicaiyang) [inline hw/dv/tools/xcelium/common_cov.vRefine:8]: can we break this to multiple lines for readability?
```suggestion
      <ccf-file content="
      //only collect toggle coverage on dut;
      deselect_coverage -toggle -instance tb.dut.*;
      // only collect toggle coverage on ports;
      set_toggle_portsonly;
     ...

File: hw/dv/tools/xcelium/common_cov.vRefine (hunk 2 above)
Comment (rasmus-madsen) [reply hw/dv/tools/xcelium/common_cov.vRefine:5]: ahh good catch - it is the output of the vRefine when saved through the gui.
not sure if it matters I will look into alternatives

File: hw/dv/tools/xcelium/common_cov.vRefine (hunk 3 above)
Comment (rasmus-madsen) [reply hw/dv/tools/xcelium/common_cov.vRefine:8]: this file is auto generated

Comment (weicaiyang) [reply hw/dv/tools/xcelium/common_cov.vRefine:8]: Sometimes, it still works when we do some minor modification. at least it works for VCS cov exclusion file
It's good to remove IP specific content and try to run with another block before we add it into common area.

**Comment by rasmus-madsen** [comment]
@sriyerg @weicaiyang 
This is still not 100%

the Cadence way of handling coverage is a little different than VCS.
as you both saw the refine files have the user file path in them.

so instead I turned to tcl files - but these have two issues.
1. only one file is allowed.
2. the instance name is actually the module name - which makes it module-specific. 
(i.e -inst  aes vs -inst dut)

I have reached out to our Cadence AE to see if he has a solution..

Another solution would be to use the refines with the suggested edits by @weicaiyang manually removing anything user-specific.
but that would mean a high risk of a user forgetting this step (will be caught in review)

do you guys have any preferences/alternative solutions?

**Comment by weicaiyang** [comment]
> @sriyerg @weicaiyang
> This is still not 100%
> 
> the Cadence way of handling coverage is a little different than VCS.
> as you both saw the refine files have the user file path in them.
> 
> so instead I turned to tcl files - but these have two issues.
> 
> 1. only one file is allowed.
> 2. the instance name is actually the module name - which makes it module-specific.
>    (i.e -inst  aes vs -inst dut)
> 
> I have reached out to our Cadence AE to see if he has a solution..
> 
> Another solution would be to use the refines with the suggested edits by @weicaiyang manually removing anything user-specific.
> but that would mean a high risk of a user forgetting this step (will be caught in review)
> 
> do you guys have any preferences/alternative solutions?

either tcl or vRefine file is fine to me, as long as we can have a common coverage file for all IPs that run with Xcelium.
if it's tcl, probably we can use env variable $DUT_TOP for the module name. For `only one file is allowed`, VCS also only allows one cover cfg file. I think we just need one common file for all IPs. if some IP need a special one, then create a new one and point to that.
If vRefine, perhaps remove module specific and try if it works with other modules.

**Comment by rasmus-madsen** [comment]
@weicaiyang if Vcs also only allow one exclude file, how do you handle the common exclusion file.
I think if it's ok to only have one exclusion file
then I just need to rename the hjson variable from "refine to exclusion" since it is not using refines.

the tcl file works better if you ask me.
only problem is you have to write it by hand. the refine you can save from the gui

**Comment by weicaiyang** [comment]
> @weicaiyang if Vcs also only allow one exclude file, how do you handle the common exclusion file.
> I think if it's ok to only have one exclusion file
> then I just need to rename the hjson variable from "refine to exclusion" since it is not using refines.
> 
> the tcl file works better if you ask me.
> only problem is you have to write it by hand. the refine you can save from the gui

sorry, I haven't use xcelium coverage flow.
For VCS, it supports one [cover cfg file](https://github.com/lowRISC/opentitan/blob/master/hw/dv/tools/vcs/cover.cfg) and multiple exclusion files. So we have a cover cfg to define the common coverage such as enabling coverage at tb.dut and only enable toggle for the dut (+tree tb.dut 1). We also have a [common exclusion file](https://github.com/lowRISC/opentitan/blob/master/hw/dv/tools/vcs/common_cov_excl.el) to exclude some TLUL ports. VCS exclusion file is also generated, but we can add/remove comment or re-org the content to make it more readable.

it will be nice if Xcelium can have this similar coverage setup: common coverage in one file, module specific in another file. perhaps use tcl file for the common and use vrefine file for module specific.

**Comment by rasmus-madsen** [comment]
> > @weicaiyang if Vcs also only allow one exclude file, how do you handle the common exclusion file.
> > I think if it's ok to only have one exclusion file
> > then I just need to rename the hjson variable from "refine to exclusion" since it is not using refines.
> > the tcl file works better if you ask me.
> > only problem is you have to write it by hand. the refine you can save from the gui
> 
> sorry, I haven't use xcelium coverage flow.
> For VCS, it supports one [cover cfg file](https://github.com/lowRISC/opentitan/blob/master/hw/dv/tools/vcs/cover.cfg) and multiple exclusion files. So we have a cover cfg to define the common coverage such as enabling coverage at tb.dut and only enable toggle for the dut (+tree tb.dut 1). We also have a [common exclusion file](https://github.com/lowRISC/opentitan/blob/master/hw/dv/tools/vcs/common_cov_excl.el) to exclude some TLUL ports. VCS exclusion file is also generated, but we can add/remove comment or re-org the content to make it more readable.
> 
> it will be nice if Xcelium can have this similar coverage setup: common coverage in one file, module specific in another file. perhaps use tcl file for the common and use vrefine file for module specific.

I already implemented the coverage configuration file ccf
with matches the vcs cfg closely.

I am going back to refine files as we can have multiple of those, and we can manually clean them up.
I reached out to cadence - and there really isn't an option to have common coverage. (which also would put strict rules on naming convention). 
IMC use module name  not instance name.
I.e AES vs DUT.


but I also think that excluding the TL signals manually is not too bad. WDYT?

**Comment by weicaiyang** [comment]
> I already implemented the coverage configuration file ccf
> with matches the vcs cfg closely.
> 
> I am going back to refine files as we can have multiple of those, and we can manually clean them up.
> I reached out to cadence - and there really isn't an option to have common coverage. (which also would put strict rules on naming convention).
> IMC use module name not instance name.
> I.e AES vs DUT.
> 
> but I also think that excluding the TL signals manually is not too bad. WDYT?

does xcelium accept one tcl file and one refine file?
If so, we can use env variable for module name, just like you did [here](https://github.com/lowRISC/opentitan/blob/master/hw/dv/tools/xcelium/cov_report.tcl#L16)
`exclude -inst $::env(DUT_TOP) -toggle 'tl_i.a_user.parity_en'`

if not, I think excluding TL ports manually is also fine.

**Comment by rasmus-madsen** [comment]
> > I already implemented the coverage configuration file ccf
> > with matches the vcs cfg closely.
> > I am going back to refine files as we can have multiple of those, and we can manually clean them up.
> > I reached out to cadence - and there really isn't an option to have common coverage. (which also would put strict rules on naming convention).
> > IMC use module name not instance name.
> > I.e AES vs DUT.
> > but I also think that excluding the TL signals manually is not too bad. WDYT?
> 
> does xcelium accept one tcl file and one refine file?
> If so, we can use env variable for module name, just like you did [here](https://github.com/lowRISC/opentitan/blob/master/hw/dv/tools/xcelium/cov_report.tcl#L16)
> `exclude -inst $::env(DUT_TOP) -toggle 'tl_i.a_user.parity_en'`
> 
> if not, I think excluding TL ports manually is also fine.

ahh yes that is actually a good Idea.
it does support one tcl file and then as many Vrefines as you want. 
let me try the substitution right away

**Comment by rasmus-madsen** [comment]
@weicaiyang @sriyerg 
I think it is finally where we want it.
@sriyerg might want to look at not using a tcl file for common coverage.
I didn't see a way around that.
and in reviews we will need to tell people to remove path specific stuff from the refines, leaving it does not affect coverage for others as it is ignored but I don't think we should keep that around

**Review by weicaiyang — APPROVED** [review]
LGTM, thanks

**Comment by rasmus-madsen** [comment]
@sriyerg are you ok with this?

File: hw/dv/tools/xcelium/xcelium.ccf [hunk 4]
```diff
@@ -0,0 +1,15 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+//only collect toggle coverage on dut
```
Comment (sriyerg) [inline hw/dv/tools/xcelium/xcelium.ccf:5]: nit: space after `//`

File: hw/dv/tools/xcelium/xcelium.ccf [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+deselect_coverage -toggle -instance tb.dut.*...
+
+// only collect toggle coverage on ports
+set_toggle_portsonly
+// enable toggle scoring of structs and multidim arrays + MDA structs
+set_toggle_scoring -sv_mda -sv_mda_of_struct
+// filter glitches
+set_toggle_strobe 0ps
+//  Filter unreachable/statically constant blocks
+set_com -log
```
Comment (sriyerg) [inline hw/dv/tools/xcelium/xcelium.ccf:15]: nit: add newline

File: hw/dv/tools/dvsim/common_sim_cfg.hjson [hunk 6]
```diff
@@ -140,8 +140,9 @@
   cover_reg_top_vcs_cov_cfg_file: "-cm_hier {dv_root}/tools/vcs/cover_reg_top.cfg"
 
   // Project defaults for Xcelium
-  // xcelium_cov_cfg_file: "{{build_mode}_xcelium_cov_cfg_file}"
-  // xcelium_cov_refine_files: ["{dv_root}/tools/xcelium/common_cov.vRefine"]
+  xcelium_cov_cfg_file: "{dv_root}/tools/xcelium/xcelium.ccf"
+  xcelium_cov_refine_files: []
+  xcelium_common_excl_file: ["{dv_root}/tools/xcelium/exclude.tcl"]
```
Comment (sriyerg) [inline hw/dv/tools/dvsim/common_sim_cfg.hjson:145]: Did you forget to check in this file?

**Review by sriyerg — CHANGES_REQUESTED** [review]
I couldn't get to experimenting with this. We can refine this later in stages.

File: hw/dv/tools/xcelium/xcelium.ccf (hunk 4 above)
Comment (rasmus-madsen) [reply hw/dv/tools/xcelium/xcelium.ccf:5]: fixed

File: hw/dv/tools/xcelium/xcelium.ccf (hunk 5 above)
Comment (rasmus-madsen) [reply hw/dv/tools/xcelium/xcelium.ccf:15]: fixed

File: hw/dv/tools/dvsim/common_sim_cfg.hjson (hunk 6 above)
Comment (rasmus-madsen) [reply hw/dv/tools/dvsim/common_sim_cfg.hjson:145]: yes I did..
added.

File: hw/dv/tools/xcelium/xcelium.ccf (hunk 5 above)
Comment (sriyerg) [reply hw/dv/tools/xcelium/xcelium.ccf:18]: Newline still appears to be needed (the red symbol is still there).

File: hw/dv/tools/xcelium/exclude.tcl [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+exclude -inst $::env(DUT_TOP) -toggle 'tl_i.a_user.parity'
+exclude -inst $::env(DUT_TOP) -toggle 'tl_i.a_user.parity_en'
+exclude -inst $::env(DUT_TOP) -toggle 'tl_i.a_param'
+exclude -inst $::env(DUT_TOP) -toggle 'tl_o.d_param'
+exclude -inst $::env(DUT_TOP) -toggle 'tl_o.d_sink'
+exclude -inst $::env(DUT_TOP) -toggle 'tl_o.d_user'
+exclude -inst $::env(DUT_TOP) -toggle 'tl_i.a_user.rsvd1'
+
```
Comment (sriyerg) [inline hw/dv/tools/xcelium/exclude.tcl:12]: nit: remove extra newline.

**Review by sriyerg — APPROVED** [review]
LGTM, thanks! Will see if we can improve upon this later.

File: hw/dv/tools/xcelium/exclude.tcl (hunk 7 above)
Comment (rasmus-madsen) [reply hw/dv/tools/xcelium/exclude.tcl:12]: fixed

File: hw/dv/tools/xcelium/xcelium.ccf (hunk 5 above)
Comment (rasmus-madsen) [reply hw/dv/tools/xcelium/xcelium.ccf:18]: hmm never seen this before - works in local flow.
anyway I added the newline

File: hw/dv/tools/xcelium/exclude.tcl [hunk 8]
```diff
@@ -0,0 +1,11 @@
+// Copyright lowRISC contributors.
```
Comment (sriyerg) [inline hw/dv/tools/xcelium/exclude.tcl:1]: I think comments in tcl file start with #. CI is complaining about the same thing.

Comment (rasmus-madsen) [reply hw/dv/tools/xcelium/exclude.tcl:1]: oh wierd that xcelium is accepting this.
Thank you

**Pull request merged**